addon/record-data.js

```diff
@@ -375,170 +485,366 @@ export default class FragmentRecordData extends RecordData {
-            if (frag._internalModel) {
-              frag._internalModel._recordData.commitWasRejected();
-            }
-          });
-        } else {
-          fragment._internalModel._recordData.commitWasRejected();
-        }
+      const inFlight = this._inFlightFragments[key];
+      if (inFlight != null) {
+        behavior.unload(inFlight);
       }
+      const fragmentData = this._fragmentData[key];
+      if (fragmentData != null) {
+        behavior.unload(fragmentData);
+      }
+      this._fragmentArrayCache[key]?.destroy();
     }
-    return super.commitWasRejected(...arguments);
+    super.unloadRecord();
   }
 
-  setAttr(key, value) {
-    return super.setAttr(key, value);
+  /**
+   * When a fragment becomes dirty, update the dirty state of the fragment's owner
+   */
+  fragmentDidDirty() {
+    assert('Fragment is not dirty', this.hasChangedAttributes());
+    if (!this._fragmentOwner) {
+      return;
+    }
+    const { recordData: owner, key } = this._fragmentOwner;
+    if (owner.isFragmentDirty(key)) {
+      // fragment owner is already dirty
+      return;
+    }
+
+    owner._fragments[key] = owner._fragmentData[key];
+    owner.notifyStateChange(key);
+    owner.fragmentDidDirty();
   }
 
-  getAttr(key) {
-    return super.getAttr(key);
+  /**
+   * When a fragment becomes clean, update the fragment owner
+   */
+  fragmentDidReset() {
+    if (!this._fragmentOwner) {
+      return;
+    }
+    const { recordData: owner, key } = this._fragmentOwner;
+    if (!owner.isFragmentDirty(key)) {
+      // fragment owner is already clean
+      return;
+    }
+
+    const behavior = owner._fragmentBehavior[key];
+    const value = owner.getFragment(key);
+    const originalValue = owner._fragmentData[key];
+    const isDirty = behavior.isDirty(value, originalValue);
+
+    if (isDirty) {
+      // fragment is still dirty
+      return;
+    }
+
+    delete owner._fragments[key];
+    owner.notifyStateChange(key);
+    owner.fragmentDidReset();
   }
 
-  hasAttr(key) {
-    return super.hasAttr(key);
+  notifyStateChange(key) {
+    this.storeWrapper.notifyStateChange(this.modelName, this.id, this.clientId, key);
   }
 
-  didCreateLocally() {
-    return super.didCreateLocally();
+  /*
+   * Ensure that any changes to the fragment record-data also update the InternalModel's
+   * state machine and fire property change notifications on the Record
+   */
+
+  _fragmentPushData(data) {
+    internalModelFor(this).setupData(data);
+  }
+  _fragmentWillCommit() {
+    internalModelFor(this).adapterWillCommit();
+  }
+  _fragmentDidCommit(data) {
+    internalModelFor(this).adapterDidCommit(data);
+  }
+  _fragmentRollbackAttributes() {
+    internalModelFor(this).rollbackAttributes();
+  }
+  _fragmentCommitWasRejected() {
+    internalModelFor(this).adapterDidInvalidate();
+  }
+  _fragmentUnloadRecord() {
+    internalModelFor(this).unloadRecord();
```

Comment:
This all exists to keep the InternalModel state in sync with RecordData. Maybe that's not necessary any more?

Reply:
Shouldn't be? There's no longer a state machine and IM has held no data directly for along time

Reply:
I've been testing against E-D 3.28.3. Are there more changes since then?

Reply:
@dwickern the big flood comes in 4.5, basically the internals are fully rewritten at that point with another big rewrite to them in the works

Reply:
I tested a bit more against E-D 2.8 and 4.5. There's more than the InternalModel's state-- InternalModel is also responsible for notifying that the record attributes changed.

Take rollback for example. The InternalModel calls `InternalModel.notifyAttributes` with the return value from `RecordData.rollbackAttributes`. When we roll back nested fragments, we also need to fire notifications for those fragments' changed attributes.